sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobClient.java

```diff
@@ -167,24 +168,41 @@ public void upload(InputStream data, long length, boolean overwrite) {
     public void uploadWithResponse(InputStream data, long length, ParallelTransferOptions parallelTransferOptions,
         BlobHttpHeaders headers, Map<String, String> metadata, AccessTier tier, BlobRequestConditions requestConditions,
         Duration timeout, Context context) {
-        BlockBlobClient blockBlobClient = this.getBlockBlobClient();
         final ParallelTransferOptions validatedParallelTransferOptions =
             ModelHelper.populateAndApplyDefaults(parallelTransferOptions);
-        if (length < validatedParallelTransferOptions.getMaxSingleUploadSize()) {
-            blockBlobClient.uploadWithResponse(data, length, headers, metadata, tier, null, requestConditions,
-                timeout, context);
-        } else {
-            BlobOutputStream blobOutputStream = BlobOutputStream.blockBlobOutputStream(client,
-                validatedParallelTransferOptions, headers, metadata, tier, requestConditions);
-            try {
-                StorageImplUtils.copyToOutputStream(data, length, blobOutputStream);
-                blobOutputStream.close();
-            } catch (IOException e) {
-                e.printStackTrace();
+        BlobOutputStream blobOutputStream = getBlobOutputStreamInternal(validatedParallelTransferOptions, headers,
+            metadata, tier, requestConditions);
+        try {
+            StorageImplUtils.copyToOutputStream(data, length, blobOutputStream);
+            blobOutputStream.close();
+        } catch (IOException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof BlobStorageException) {
+                throw logger.logExceptionAsError((BlobStorageException) cause);
+            } else {
+                throw logger.logExceptionAsError(new UncheckedIOException(e));
             }
         }
     }
 
+    /**
+     * An abstraction that allows child packages to inject the correct client into the blob outputstream used to upload.
+     *
+     * @param parallelTransferOptions The options passed to upload. Pre-verified.
+     * @param headers The headers passed to upload.
+     * @param metadata The metadata passed to upload.
+     * @param tier The tier passed to upload.
+     * @param requestConditions The request conditions passed to upload.
+     *
+     * @return The {@link BlobOutputStream} initialized with the appropriate client.
+     */
+    protected BlobOutputStream getBlobOutputStreamInternal(ParallelTransferOptions parallelTransferOptions,
```

Comment:
Can this be private instead? You can declare another private method in `EncryptedBlobClient` instead of overriding.